client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -211,7 +211,7 @@ private void failConnection(Throwable throwable) {
                 }
                 log.info("Handling exception {} for connection {} on writer {}. SetupCompleted: {}, Closed: {}",
                          throwable, connection, writerId, connectionSetupCompleted == null ? null : connectionSetupCompleted.isDone(), closed);
-                if (exception == null) {
+                if (exception == null || throwable instanceof RetriesExhaustedException) {
```

Comment:
I suggest using `Exceptions.unwrap` here in case the exception is wrapped in a `CompletionException` or `ExecutionException`.

Reply:
Same below

Reply:
done. this will ensure `state.exception ` is always unwrapped.